sql: refactor `checkScanParallelizationIfLocal`

mgartner opened the pull request:
#### sql: replace `walkPlan` in `checkScanParallelizationIfLocal`

`checkScanParallelizationIfLocal` now traverses a `planNode` tree using
the `Input` and `InputCount` methods instead of `walkPlan`.

Epic: None
Release note: None

#### sql: fix checkScanParallelizationIfLocal for groupNode

This commit fixes a bug in `checkScanParallelizationIfLocal` that caused
it to ignore all but the last aggregate function in a `groupNode`. This
has been fixed and a test has been added. In addition, an early return
has been added to the `renderNode` case.

Release note: None

mgartner committed:
sql: replace `walkPlan` in `checkScanParallelizationIfLocal`

`checkScanParallelizationIfLocal` now traverses a `planNode` tree using
the `Input` and `InputCount` methods instead of `walkPlan`.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/138692)
<!-- Reviewable:end -->

mgartner commented:
```
name                                              old time/op    new time/op    delta
EndToEnd/kv-read-const/vectorize=on/Prepared-16      146µs ± 1%     145µs ± 1%  -0.64%  (p=0.000 n=20+19)
EndToEnd/kv-read-const/vectorize=on/Simple-16        161µs ± 1%     161µs ± 1%  -0.57%  (p=0.007 n=20+20)
EndToEnd/kv-read/vectorize=off/Simple-16             207µs ± 1%     206µs ± 1%  -0.54%  (p=0.000 n=19+20)
EndToEnd/kv-read/vectorize=on/Prepared-16            145µs ± 1%     145µs ± 1%  -0.41%  (p=0.000 n=19+19)
EndToEnd/kv-read-const/vectorize=off/Prepared-16     139µs ± 1%     138µs ± 1%  -0.40%  (p=0.028 n=20+20)
EndToEnd/kv-read/vectorize=on/Simple-16              213µs ± 1%     212µs ± 1%    ~     (p=0.052 n=20+20)
EndToEnd/kv-read/vectorize=off/Prepared-16           139µs ± 1%     139µs ± 2%    ~     (p=0.149 n=19+20)
EndToEnd/kv-read-const/vectorize=off/Simple-16       155µs ± 1%     155µs ± 2%    ~     (p=0.351 n=19+20)

name                                              old alloc/op   new alloc/op   delta
EndToEnd/kv-read-const/vectorize=off/Simple-16      22.3kB ± 2%    22.1kB ± 1%  -0.77%  (p=0.001 n=20+16)
EndToEnd/kv-read-const/vectorize=off/Prepared-16    21.6kB ± 0%    21.6kB ± 0%  -0.23%  (p=0.023 n=17+18)
EndToEnd/kv-read/vectorize=off/Simple-16            33.8kB ± 0%    33.8kB ± 0%  -0.15%  (p=0.000 n=20+19)
EndToEnd/kv-read/vectorize=on/Simple-16             33.4kB ± 0%    33.4kB ± 0%    ~     (p=0.110 n=20+20)
EndToEnd/kv-read/vectorize=on/Prepared-16           20.5kB ± 0%    20.5kB ± 0%    ~     (p=0.151 n=20+19)
EndToEnd/kv-read/vectorize=off/Prepared-16          21.7kB ± 0%    21.7kB ± 0%    ~     (p=0.914 n=19+19)
EndToEnd/kv-read-const/vectorize=on/Simple-16       23.5kB ± 1%    23.5kB ± 1%    ~     (p=0.147 n=19+20)
EndToEnd/kv-read-const/vectorize=on/Prepared-16     20.4kB ± 2%    20.4kB ± 2%    ~     (p=0.372 n=20+20)

name                                              old allocs/op  new allocs/op  delta
EndToEnd/kv-read/vectorize=on/Prepared-16              186 ± 0%       184 ± 0%  -0.70%  (p=0.000 n=20+20)
EndToEnd/kv-read-const/vectorize=off/Simple-16         193 ± 0%       192 ± 0%  -0.69%  (p=0.000 n=18+16)
EndToEnd/kv-read-const/vectorize=off/Prepared-16       175 ± 0%       173 ± 0%  -0.69%  (p=0.000 n=17+18)
EndToEnd/kv-read-const/vectorize=on/Prepared-16        182 ± 0%       180 ± 0%  -0.63%  (p=0.000 n=20+20)
EndToEnd/kv-read-const/vectorize=on/Simple-16          217 ± 1%       216 ± 1%  -0.49%  (p=0.000 n=20+19)
EndToEnd/kv-read/vectorize=off/Prepared-16             178 ± 0%       178 ± 0%  -0.43%  (p=0.000 n=19+20)
EndToEnd/kv-read/vectorize=off/Simple-16               267 ± 0%       266 ± 0%  -0.37%  (p=0.000 n=20+19)
EndToEnd/kv-read/vectorize=on/Simple-16                280 ± 0%       278 ± 0%  -0.36%  (p=0.000 n=20+20)
```

```
name                                         old time/op    new time/op    delta
Sysbench/SQL/1node_local/oltp_read_write-16    4.64ms ± 1%    4.61ms ± 1%  -0.51%  (p=0.015 n=20+20)
Sysbench/SQL/1node_local/oltp_read_only-16     2.69ms ± 1%    2.68ms ± 2%    ~     (p=0.289 n=20+20)

name                                         old alloc/op   new alloc/op   delta
Sysbench/SQL/1node_local/oltp_read_only-16      804kB ± 0%     804kB ± 0%    ~     (p=0.081 n=20+20)
Sysbench/SQL/1node_local/oltp_read_write-16    1.36MB ± 0%    1.36MB ± 0%    ~     (p=0.384 n=18+16)

name                                         old allocs/op  new allocs/op  delta
Sysbench/SQL/1node_local/oltp_read_only-16      3.78k ± 0%     3.77k ± 0%  -0.23%  (p=0.014 n=20+20)
Sysbench/SQL/1node_local/oltp_read_write-16     6.50k ± 0%     6.49k ± 1%    ~     (p=0.064 n=20+20)
```

mgartner commented on pkg/sql/distsql_physical_planner.go:
```diff
@@ -5021,82 +5021,71 @@ func checkScanParallelizationIfLocal(
+			prohibitParallelization = true
+		case *groupNode:
+			for _, f := range n.funcs {
+				prohibitParallelization = f.hasFilter()
```
A few of these cases that unconditionally set `prohibitParallelization` seem incorrect to me, but I've left the logic as-is for now (or at least I think I have).

yuzefovich replied:
No, this is by design - filtering aggregation is not supported natively in the vectorized engine (see `errFilteringAggregation` in `execplan.go`), so if we have a `groupNode` with a filter set, then we will wrap row-by-row aggregator to handle it, and as a result we cannot parallelize the scans in the whole flow if local (see comment on `checkScanParallelizationIfLocal`).
<!-- Reviewable comment -OG7TApl6H-FSkt-ZVL3:b-5vpyq6 -->
<!-- Sent from Reviewable.io -->

mgartner replied:
This loop is equivalent to:

```go
prohibitParallelization = n.funcs[len(n.funcs)-1].hasFilter()
```

Only the last of `funcs` affects `prohibitParallelization`. Is that correct?

yuzefovich replied:
Oh, I see what you mean. Yeah, this was a bug. Nice catch! Perhaps add a regression test to `TestCheckScanParallelizationIfLocal`?

For `joinNode` case I think the code worked but in a tricky way. We always checked `c.prohibitParallelization` at the top of the method, so if we got to `joinNode` case, then `c.prohibitParallelization` must have been `false` initially; then if the join node has the ON expression, we do want to prohibit the parallelization going forward. Still, it'd be cleaner to conditionally update `prohibitParallelization` in this case too.
<!-- Reviewable comment -OGB0tsD5PwXGgtj87xG:bqjui3v -->
<!-- Sent from Reviewable.io -->

mgartner replied:
I've fixed the `groupNode` bug and added a test for it.

mgartner commented on pkg/sql/distsql_physical_planner.go:
```diff
@@ -5021,82 +5021,71 @@ func checkScanParallelizationIfLocal(
+			}
+			recurse = true
+		case *joinNode:
+			prohibitParallelization = n.pred.onCond != nil
```
Is this incorrect?

yuzefovich replied:
We only natively support the ON expression on hash and merge joins for INNER join type in the vectorized engine, so here we're prohibiting the scan parallelization in a bit more cases than necessary.

We also need to add `recurse = true` for `joinNode` case. I wonder whether we should default `recurse` to `true` above since we have fewer cases when we don't recurse. Alternatively, we could remove `recurse` altogether in favor of short-circuiting via `return` in each case when we don't want to explore further.
<!-- Reviewable comment -OG7TlpPE0rXnl38SLXs:bn5ast8 -->
<!-- Sent from Reviewable.io -->

mgartner replied:
Good idea re: `recurse`, done.
<!-- Reviewable comment -OGAYzz-2vXbXeUFE0qG:bm9lisb -->
<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Nice! I have one suggestion.

> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138692#-:-OG7UmGT0QtQKiIu0XJa:bctmxe8)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[`pkg/sql/distsql_physical_planner.go` line 5038 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/138692#-OG7SvGsDK5MOjwKzsnS:-OG7SvGsDK5MOjwKzsnT:b-nhv8qu) ([raw file](https://github.com/cockroachdb/cockroach/blob/022b2cbdaf704a15b9ccd1a772a262823bb88382/pkg/sql/distsql_physical_planner.go#L5038)):*
> ```Go
> 			// explainPlanNode is a zeroInputPlanNode, so we have to manually
> 			// recurse.
> 			// walkPlan doesn't recurse into explainPlanNode, so we have to manually
> ```

nit: leftover comment.


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138692)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @rytaft and @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> Reviewed 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138692)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)



<!-- Sent from Reviewable.io -->

mgartner committed:
sql: fix checkScanParallelizationIfLocal for groupNode

This commit fixes a bug in `checkScanParallelizationIfLocal` that caused
it to ignore all but the last aggregate function in a `groupNode`. This
has been fixed and a test has been added. In addition, an early return
has been added to the `renderNode` case.

Release note: None

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 3 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/138692#-:-OMF-5gHAszVJHthJhfT:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @rytaft)



<!-- Sent from Reviewable.io -->

mgartner commented:
TFTRs!

bors r+
